i2c/common/libi2c-msg.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Phoenix-RTOS
+ *
+ * i2c msg client library
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Marek Bialowas
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <sys/msg.h>
+#include <sys/types.h>
+
+#include <stdio.h>
+#include <errno.h>
+
+#include <i2c.h>
+#include <i2c-msg.h>
+
+static struct {
+	oid_t i2c_oid;
+	int initialized;
+} common;
+
+/* opens the device */
+int i2c_init(unsigned int dev_no)
+{
+	char devname[sizeof("/dev/i2cX")];
+	int res = 0;
+
+	sprintf(devname, "/dev/i2c%u", dev_no % 10);
+	res = lookup(devname, NULL, &common.i2c_oid);
+
+	if (res >= 0)
+		common.initialized = 1;
+
+	return res;
+}
+
+
+/* Performs i2c generic write operation to the given slave device. */
+int i2c_busWrite(uint8_t dev_addr, const uint8_t *data, uint32_t len)
+{
+	msg_t msg;
+	int res;
+	i2c_devctl_t *i = (i2c_devctl_t *)msg.i.raw;
+
+	if (!common.initialized)
+		return -EIO;
+
+	msg.type = mtDevCtl;
+
+	msg.i.data = (uint8_t *)data; /* FIXME: dropping const because of broken msg_t declaration */
+	msg.i.size = len;
+	msg.o.data = NULL;
+	msg.o.size = 0;
+
+	i->type = i2c_devctl_bus_write;
+	i->dev_addr = dev_addr;
+
+	if ((res = msgSend(common.i2c_oid.port, &msg)) < 0)
+		return res;
+
+	return msg.o.io.err;
+}
+
+
+/* Performs i2c generic read operation from the given slave device. */
+int i2c_busRead(uint8_t dev_addr, uint8_t *data_out, uint32_t len)
+{
+	msg_t msg;
+	int res;
+	i2c_devctl_t *i = (i2c_devctl_t *)msg.i.raw;
+
+	if (!common.initialized)
+		return -EIO;
+
+	msg.type = mtDevCtl;
+
+	msg.i.data = NULL;
+	msg.i.size = 0;
+	msg.o.data = data_out;
+	msg.o.size = len;
+
+	i->type = i2c_devctl_bus_read;
+	i->dev_addr = dev_addr;
+
+	if ((res = msgSend(common.i2c_oid.port, &msg)) < 0)
+		return res;
+
+	return msg.o.io.err;
```

Comment:
As above, this field is for mtOpen, mtRead etc, but not for mtDevCtl.

Reply:
hehh, seems for mtCreate I need to use `msg.o.create.err` (the and other errors for `mtLookup`, and `mtGetAttr`, `mtSetAttr` ...) :P